### DISCUSSION TITLE
Server-side-streaming: close receiver on client and detect on server

### DISCUSSION THREAD

**Discussion by g0hl1n** [hyperium/tonic#1190] [Q&A]
Hi,
I'm currently trying to implement a test for my gRPC server-side-streaming rpc. As this is some kind of "event" channel it is quite long-lived and clients may close their receiver side. Additionally I need to keep track of how many clients currently are receiving the stream.

When a client from Java code closes/drops the receiver channel the following server side code works as expected:

```rust
let (tx, rx) = mpsc::channel::<Result<SssResponse, Status>>(1);
tokio::spawn(async move {
    loop {
        select! {
            () = tx.closed() => {
                debug!("client closed server side stream");
                break;
            },
            a = receiver.recv() => match a {
                // do some things and then tx.send()
            },
        }
    }
}
```

When trying to close the receiver with rust I don't get the tx.closed() future. This is the code I'm using to close the stream in my test:

```rust
let mut stream1 = si.server_side_stream().await.unwrap().into_inner();
stream1.close();
while stream1.as_mut().recv().await.is_some() {}
drop(stream1);
```

Do you have any ideas why I don't get that tx.closed() with the stream drop code from above?

Any help is greatly appreciated! Thanks!

**Comment by LucioFranco** [comment]
This seems like more related to why the channel is having issues than tonic? Without more information on the types used and what server_side_stream returns its hard to say without seeing more code.
